File: Heap/Leetcode/Hard/MergeKLists/MergeKLists.cpp
// https://leetcode.com/problems/merge-k-sorted-lists/
#include <bits/stdc++.h>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class compare
{
public:
    bool operator()(ListNode *a, ListNode *b)
    {
        return a->val > b->val;
    }
};


ListNode *mergeKLists(vector<ListNode *> &lists)
{
    priority_queue<ListNode *, vector<ListNode *>, compare> pq;

    for (int i = 0; i < lists.size(); i++)
    {
        if (lists[i])
        {
            pq.push(lists[i]);
        }
    }
    if (pq.empty())
        return nullptr;
    ListNode *head = pq.top(), *ptr = head;
    pq.pop();
    if (ptr->next != nullptr)
    {
        pq.push(ptr->next);
    }

    while (!pq.empty())
    {
        ptr->next = pq.top();
        pq.pop();
        ptr = ptr->next;
        if (ptr->next)
        {
            pq.push(ptr->next);
        }
    }

    return head;
}
int main(void)
{

    return 0;
}

File: Heap/Leetcode/Hard/MergeKLists/MergeKLists.java
// https://leetcode.com/problems/merge-k-sorted-lists/

import java.util.PriorityQueue;

class ListNode {

    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

}

public class MergeKLists {

    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);

        for (int i = 0; i < lists.length; i++) {
            if (lists[i] != null) {
                pq.offer(lists[i]);
            }
        }
        if (pq.isEmpty()) {
            return null;
        }
        ListNode head = pq.poll(), ptr = head;
        if (ptr.next != null) {
            pq.offer(ptr.next);
        }
        while (!pq.isEmpty()) {
            ptr.next = pq.poll();
            ptr = ptr.next;
            if (ptr.next != null) {
                pq.offer(ptr.next);
            }
        }

        return head;
    }
}
